### PR TITLE
Update `run_inference_algorithm` to split `initial_position` and `initial_state`

### PR DESCRIPTION
**Pull request by reubenharry**
Improvement to `run_inference_algorithm`

**Description**:

`run_inference_algorithm` currently uses a try-except clause to allow the user to either provide an initial position or an initial state. This has led to some problems, when the `except` clause fails to trigger.

`run_inference_algorithm` also produces an array of `[num_steps, n_dims]`, which for long chains on high-dim problems can be prohibitively large. 

**Solution**:

Make `run_inference_algorithm` take `initial_state` or `initial_position` explicitly, so that the caller of `run_inference_algorithm` is responsible for providing `initial_state` or `initial_position`, rather than deferring this task.

Add a `streaming` mode for `run_inference_algorithm`.

 A few important guidelines and requirements before we can merge your PR:

 - [x] The branch is rebased on the latest `main` commit;
 - [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [x] The code respects the current naming conventions;
 - [x] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [x] There are tests covering the changes;
 - [x] The doc is up-to-date;

### CODE REVIEW

File: blackjax/util.py [hunk 1]
```diff
@@ -142,12 +142,13 @@ def index_pytree(input_pytree: ArrayLikeTree) -> ArrayTree:
 
 def run_inference_algorithm(
     rng_key: PRNGKey,
-    initial_state_or_position: ArrayLikeTree,
+    initial_state: ArrayLikeTree,
```
Comment (junpenglao) [inline blackjax/util.py:145]: While I agree the initial `try ... except ...` block is not rigor to capture both `initial_state` and `initial_position`, we should not remove it here because an important use case for this top level API is for user to just supply the `initial_position`. Let's do it this way:

```python
def run_inference_algorithm(
    rng_key: PRNGKey,
    inference_algorithm: Union[SamplingAlgorithm, VIAlgorithm],
    num_steps: int,
    initial_state: ArrayLikeTree = None,
    initial_position: ArrayLikeTree = None,
    progress_bar: bool = False,
    transform: Callable = lambda x: x,
):
    ...
    if initial_state is None and initial_position is None:
        raise ...
    if initial_state is not None and initial_position is not None:
        raise ...
    if initial_position is not None:
        rng_key, init_key = split(rng_key, 2)
        inital_state = inference_algorithm.init(initial_state_or_position, init_key)
```

File: blackjax/util.py [hunk 2]
```diff
@@ -142,12 +142,13 @@ def index_pytree(input_pytree: ArrayLikeTree) -> ArrayTree:
...
     inference_algorithm: Union[SamplingAlgorithm, VIAlgorithm],
     num_steps: int,
     progress_bar: bool = False,
     transform: Callable = lambda x: x,
-) -> tuple[State, State, Info]:
+    streaming=False,
```
Comment (junpenglao) [inline blackjax/util.py:150]:
```suggestion
    return_states_avg=False,
```

File: blackjax/util.py [hunk 3]
```diff
@@ -171,6 +171,8 @@ def run_inference_algorithm(
         A transformation of the trace of states to be returned. This is useful for
         computing determinstic variables, or returning a subset of the states.
         By default, the states are returned as is.
+    streaming
+        if True, `run_inference_algorithm` will take a streaming average of the value of transform, and return that average instead of the full set of samples. This is useful when memory is a bottleneck.
```
Comment (junpenglao) [inline blackjax/util.py:175]: also no diagnostic is return when set to True

File: blackjax/util.py [hunk 4]
```diff
@@ -179,26 +181,64 @@ def run_inference_algorithm(
...
-    keys = split(sample_key, num_steps)
+    keys = split(rng_key, num_steps)
 
     @jit
     def _one_step(state, xs):
         _, rng_key = xs
         state, info = inference_algorithm.step(rng_key, state)
         return state, (transform(state), info)
 
+    def _online_one_step(average_and_state, xs):
```
Comment (junpenglao) [inline blackjax/util.py:193]: There is too much code repetition here, I suggest always computing the average as additional state for now.

**Review by junpenglao — CHANGES_REQUESTED** [review]

File: blackjax/util.py (hunk 1 above)
Comment (reubenharry) [reply blackjax/util.py:145]: OK, I'm open to that. But I'd argue that this isn't ideal. The user has access to `inference_algorithm.init`, so can easily produce `initial_state` from `initial_position` in just one line. Hiding this process inside `run_inference_algorithm` doesn't really make it easier for the user, and creates an awkward `initial_state_or_position` variable along with various errors that have to be raised when both or neither are present. 

tldr: benefit of this feature is very low, added code complexity is high

File: blackjax/util.py (hunk 2 above)
Comment (reubenharry) [reply blackjax/util.py]: I did `return_expectation`, is that OK?

File: blackjax/util.py (hunk 3 above)
Comment (reubenharry) [reply blackjax/util.py]: This is true. Fixing this was slightly tricky, and I think should be deferred to an upcoming change which will change transform to also take info. I didn't want to do this here because it breaks other parts of the codebase.

File: blackjax/util.py (hunk 4 above)
Comment (reubenharry) [reply blackjax/util.py]: Addressed.

File: blackjax/util.py (hunk 1 above)
Comment (junpenglao) [reply blackjax/util.py:146]: It is a valid point, but then maybe we should change to require the input of `inference_algorithm.step` in the API?

Comment (reubenharry) [reply blackjax/util.py:146]: That seems reasonable. If you like, I can do that now.

Comment (reubenharry) [reply blackjax/util.py]: OK, I put it in, as you outlined. This required changing the order of arguments, so I went through the code and made sure `run_inference_algorithm` is called with default arguments everywhere.

**Review by junpenglao — APPROVED** [review]

File: blackjax/util.py [hunk 5]
```diff
@@ -142,12 +143,15 @@ def index_pytree(input_pytree: ArrayLikeTree) -> ArrayTree:
...
-    initial_state_or_position: ArrayLikeTree,
     inference_algorithm: Union[SamplingAlgorithm, VIAlgorithm],
     num_steps: int,
+    initial_state: ArrayLikeTree = None,
+    initial_position: ArrayLikeTree = None,
     progress_bar: bool = False,
     transform: Callable = lambda x: x,
-) -> tuple[State, State, Info]:
+    return_state_history=True,
+    expectation: Callable = lambda x: x,
```
Comment (junpenglao) [inline blackjax/util.py:153]: Should we combine the kwargs of `transform` and `expectation` for now? Otherwise we need a better name for `expectation`

File: blackjax/util.py [hunk 6]
```diff
@@ -171,34 +176,88 @@ def run_inference_algorithm(
...
+    )
+
+    if not return_state_history:
+        return average, transform(final_state)
+    else:
+        state_history, info_history = history
+        return transform(final_state), state_history, info_history
+
+
+def streaming_average(O, x, streaming_avg, weight=1.0, zero_prevention=0.0):
```
Comment (junpenglao) [inline blackjax/util.py:235]: we need a more descriptive name here instead of `O`. Moreover, I think it is better to have only `x` as input and have O(x) being called in `one_step` above.

**Review by junpenglao — CHANGES_REQUESTED** [review]
Actually, a few more nit

File: blackjax/util.py (hunk 5 above)
Comment (reubenharry) [reply blackjax/util.py:153]: `expectation` seems to me like an appropriate name, since indeed the value calculated is the expectation. For example, if `expectation=lambda x: x**2`, then you get back $E[x^2]$.

By contrast, `transform` operates on the full history of samples. In the future, I think it would make sense for it to also take Info, so that the user can choose to dispense with (part of) the diagnostic info. That is why I thought it was better to keep them separate.

I think it would in theory be possible to make them the same, if necessary.

File: blackjax/util.py (hunk 6 above)
Comment (reubenharry) [reply blackjax/util.py]: Done. I used `expectation` as the new name.

File: blackjax/util.py [hunk 7]
```diff
@@ -171,34 +176,87 @@ def run_inference_algorithm(
...
+
+
+def streaming_average(expectation, streaming_avg, weight=1.0, zero_prevention=0.0):
+    """Compute the streaming average of a function O(x) using a weight.
+    Parameters:
+    ----------
+        O
+            function to be averaged
+        x
+            current state
```
Comment (junpenglao) [inline blackjax/util.py:242]: Update docstring

File: blackjax/util.py (hunk 5 above)
Comment (junpenglao) [reply blackjax/util.py:153]: That makes sense, but looking at the implementation, the return state is always `transform(state)`, which means for a expectation transformation, we should probably do `expectation(transform(x))`?

**Comment by reubenharry** [comment]
The other argument returned, `average`, is the expectation. I'm not sure `expectation(transform(x))` would work, because they are both functions of `x` (the state).

**Comment by junpenglao** [comment]
> The other argument returned, average, is the expectation. I'm not sure expectation(transform(x)) would work, because they are both functions of x (the state).

What i meant here is that:
1. The function intially returns `transform(x)`, which is a function of x (the state). For examples with transform is `lambda x: x.position`. If user wants to compute expectation eg $E[x^2]$, they need to do `np.mean(mcmc_samples ** 2)`
2. Now with the new functionarity that computes the expectation and not storing the full trace, my arguement here is that we should have `trasnform` being a function of the state, but function `expectation` being a function of the output of `transform`.

Does this makes sense?

**Comment by reubenharry** [comment]
Yeah, that makes sense. Updated.

**Review by junpenglao — APPROVED** [review]
Thank you!
